File: experiments/001_wow_fishing/record.py
"""Record one small, silent macOS screen region; no model calls or input events."""

import argparse
from datetime import datetime, timezone
import json
from pathlib import Path
import subprocess
import sys
import uuid


def rectangle(value):
    try:
        x, y, width, height = map(int, value.split(","))
    except ValueError as exc:
        raise argparse.ArgumentTypeError("Use x,y,width,height in desktop coordinates") from exc
    if not (0 < width <= 1000 and 0 < height <= 1000):
        raise argparse.ArgumentTypeError("Width and height must be between 1 and 1000")
    return f"{x},{y},{width},{height}"


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--rect", required=True, type=rectangle)
    parser.add_argument("--seconds", type=int, choices=range(1, 31), default=30,
                        metavar="1..30")
    parser.add_argument("--dry-run", action="store_true", help="Print settings without recording")
    args = parser.parse_args()
    if sys.platform != "darwin":
        parser.error("This recorder requires macOS")
    folder = Path(__file__).resolve().parents[2] / "data" / "001_wow_fishing"
    run_id = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ") + "_" + uuid.uuid4().hex[:8]
    output = folder / f"{run_id}.mov"
    command = ["/usr/sbin/screencapture", "-v", "-x", f"-V{args.seconds}",
               f"-R{args.rect}", str(output)]
    if args.dry_run:
        print(json.dumps({"command": command, "audio": False, "model_calls": 0}, indent=2))
        return
    folder.mkdir(parents=True, exist_ok=True)
    metadata = {"started_at_utc": datetime.now(timezone.utc).isoformat(),
                "desktop_rect": args.rect, "requested_seconds": args.seconds,
                "audio": False, "status": "started", "file": output.name}
    note = output.with_suffix(".json")
    note.write_text(json.dumps(metadata, indent=2) + "\n")
    print(f"Recording for {args.seconds} seconds: {output}", flush=True)
    try:
        subprocess.run(command, check=True, timeout=args.seconds + 15)
        if not output.exists() or output.stat().st_size == 0:
            raise RuntimeError("No recording produced; check Screen Recording permission")
    except (subprocess.SubprocessError, OSError, RuntimeError) as exc:
        metadata["status"] = "failed"
        print(f"Recording failed: {exc}", file=sys.stderr)
        code = 1
    else:
        metadata["status"] = "captured_unreviewed"
        metadata["bytes"] = output.stat().st_size
        print(f"Saved {metadata['bytes'] / 1024 / 1024:.2f} MiB; inspect before using as evidence.")
        code = 0
    metadata["finished_at_utc"] = datetime.now(timezone.utc).isoformat()
    note.write_text(json.dumps(metadata, indent=2) + "\n")
    raise SystemExit(code)


if __name__ == "__main__":
    main()

File: experiments/001_wow_fishing/record.swift
// Record one small, silent macOS screen region; no model calls or input events.
//
//   record --rect x,y,width,height [--seconds 1..30] [--dry-run]
//
// Built by build.sh with ../002_wow_visual/runtime/JSON.swift.
import Foundation

@main
struct Record {
    static func main() {
        var rect: String?, seconds = 30, dryRun = false
        var rest = CommandLine.arguments.dropFirst()
        while let flag = rest.popFirst() {
            switch flag {
            case "--rect": rect = rest.popFirst()
            case "--seconds":
                guard let value = rest.popFirst().flatMap(Int.init), (1...30).contains(value) else { usage("--seconds must be 1..30") }
                seconds = value
            case "--dry-run": dryRun = true
            default: usage("unknown argument " + flag)
            }
        }
        let parts = (rect ?? "").split(separator: ",", omittingEmptySubsequences: false).map { Int($0) }
        guard parts.count == 4, parts.allSatisfy({ $0 != nil }) else { usage("--rect: use x,y,width,height in desktop coordinates") }
        guard (1...1000).contains(parts[2]!), (1...1000).contains(parts[3]!) else { usage("--rect: width and height must be between 1 and 1000") }
        let region = parts.map { String($0!) }.joined(separator: ",")
        let folder = URL(fileURLWithPath: #filePath).deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent()
            .appendingPathComponent("data/001_wow_fishing")
        let clock = DateFormatter()
        clock.dateFormat = "yyyyMMdd'T'HHmmss'Z'"
        clock.timeZone = TimeZone(identifier: "UTC")
        clock.locale = Locale(identifier: "en_US_POSIX")
        let id = UUID().uuidString.lowercased().replacingOccurrences(of: "-", with: "").prefix(8)
        let output = folder.appendingPathComponent("\(clock.string(from: Date()))_\(id).mov")
        let command = ["/usr/sbin/screencapture", "-v", "-x", "-V\(seconds)", "-R\(region)", output.path]
        if dryRun {
            print(JSON.object([("command", .array(command.map(JSON.string))), ("audio", .bool(false)), ("model_calls", .int(0))]).text(indent: 2))
            return
        }
        do {
            try FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)
        } catch {
            fail("Recording failed: \(error)")
        }
        let note = output.deletingPathExtension().appendingPathExtension("json")
        var metadata = JSON.object([("started_at_utc", .string(utc())), ("desktop_rect", .string(region)), ("requested_seconds", .int(seconds)),
                                    ("audio", .bool(false)), ("status", .string("started")), ("file", .string(output.lastPathComponent))])
        func save() { try? (metadata.text(indent: 2) + "\n").write(to: note, atomically: true, encoding: .utf8) }
        save()
        print("Recording for \(seconds) seconds: \(output.path)")
        fflush(stdout)
        var code: Int32 = 0
        do {
            let process = Process()
            process.executableURL = URL(fileURLWithPath: command[0])
            process.arguments = Array(command.dropFirst())
            try process.run()
            let deadline = ProcessInfo.processInfo.systemUptime + Double(seconds + 15)
            while process.isRunning && ProcessInfo.processInfo.systemUptime < deadline { usleep(100_000) }
            if process.isRunning {
                process.terminate()
                process.waitUntilExit()
                throw RecordError(description: "screencapture ran past \(seconds + 15) s")
            }
            let size = (try? FileManager.default.attributesOfItem(atPath: output.path)[.size] as? Int) ?? nil
            guard process.terminationStatus == 0, let size, size > 0 else {
                throw RecordError(description: "No recording produced; check Screen Recording permission")
            }
            metadata = metadata.setting("status", .string("captured_unreviewed")).setting("bytes", .int(size))
            print(String(format: "Saved %.2f MiB; inspect before using as evidence.", Double(size) / 1024 / 1024))
        } catch {
            metadata = metadata.setting("status", .string("failed"))
            FileHandle.standardError.write(Data("Recording failed: \(error)\n".utf8))
            code = 1
        }
        metadata = metadata.setting("finished_at_utc", .string(utc()))
        save()
        exit(code)
    }

    struct RecordError: Error, CustomStringConvertible { let description: String }

    static func utc() -> String {
        let format = DateFormatter()
        format.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSSSSxxxxx"
        format.timeZone = TimeZone(identifier: "UTC")
        format.locale = Locale(identifier: "en_US_POSIX")
        return format.string(from: Date())
    }

    static func usage(_ why: String) -> Never {
        FileHandle.standardError.write(Data("usage: record --rect x,y,width,height [--seconds 1..30] [--dry-run]\nerror: \(why)\n".utf8))
        exit(2)
    }

    static func fail(_ why: String) -> Never {
        FileHandle.standardError.write(Data((why + "\n").utf8))
        exit(1)
    }
}
